village/src/App.js

```diff
@@ -11,14 +16,89 @@ class App extends Component {
       smurfs: [],
     };
   }
-  // add any needed code to ensure that the smurfs collection exists on state and it has data coming from the server
-  // Notice what your map function is looping over and returning inside of Smurfs.
-  // You'll need to make sure you have the right properties on state and pass them down to props.
+
+  componentDidMount() {
+    this.getSmurfs();
+  }
+
+  //====== API REQUEST CODE ======//
+
+  request = (url, method, successCallback, failCallback, payload) => {
+    switch (method) {
+    case 'GET':
+      axios.get(url)
+        .then(res => {
+          successCallback(res.data);
+        })
+        .catch(err => {
+          failCallback(err);
+        })
+      break;
+    case 'POST':
+      axios.post(url, payload)
+        .then(res => {
+          successCallback(res.data);
+        })
+        .catch(err => {
+          failCallback(err);
+        })
+      break;
+    case 'DELETE':
+      axios.delete(url, payload)
+        .then(res => {
+          successCallback(res.data);
+        })
+        .catch(err => {
+          failCallback(err);
+        })
+      break;
+    default:
+      console.error(`Request to ${url} failed. Please check your config.`);
+    }
+  }
+
+  getSmurfs = () => {
+    const setSmurfsToState = (data) => this.setDataToState(data, ['smurfs']);
+    this.request(apiURL, 'GET', setSmurfsToState);
+  }
+
+  addSmurf = (smurf) => {
+    const setSmurfsToState = (data) => this.setDataToState(data, ['smurfs']);
+    this.request(apiURL, 'POST', setSmurfsToState, null, smurf);
+  }
+
+  deleteSmurf = (id) => {
+    const setSmurfsToState = (data) => this.setDataToState(data, ['smurfs']);
+    this.request(`${apiURL}/${id}`, 'DELETE', setSmurfsToState);
+  }
+
+  //====== STATE UTILITY FUNCTIONS ======//
+
+  setDataToState = (data, stateKeysArray) => {
```

Comment:
This is certainly interesting piece of code. It seems to be working perfectly but i personally believe it is a little overcomplicated for simple requests to the server. It is worth remembering that the more complicated the code the harder it is to debug and test it properly when something breaks.

This is interesting approach that i will look at more closely over the weekend 👍